{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EXAM 2: ANSWERS\n",
    "\n",
    "### Fundamentals of programming and bioinformatics\n",
    "\n",
    "#### Sergio LÃ³pez Padilla"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "germplasm = open('Germplasm.tsv', 'r')\n",
    "locus = open('LocusGene.tsv', 'r')\n",
    "\n",
    "#selecting first columns\n",
    "\n",
    "AGI_germplasm = []\n",
    "AGI_locus = []\n",
    "\n",
    "with germplasm as file:\n",
    "    reader = csv.reader (file, delimiter = \"\\t\")\n",
    "    for row in reader:\n",
    "        AGI_germplasm.append(row[0]) \n",
    "        \n",
    "with locus as file:\n",
    "    reader = csv.reader (file, delimiter = \"\\t\")\n",
    "    for row in reader:\n",
    "        AGI_locus.append(row[0])\n",
    "        \n",
    "#print(AGI_germplasm)\n",
    "#print(AGI_locus)\n",
    "\n",
    "#comparing \n",
    "if AGI_germplasm == AGI_locus:\n",
    "    print(\"The list are equal\")\n",
    "else:\n",
    "    print(\"The list are not equal\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To prove that the files \"Germplasm.tsv\" and \"LocusGene.tsv\" are in the same sequence I have decided to compare the AGI code column (first one) of each file.\n",
    "\n",
    "First I have opened the two files using \"open\" command. After that, I have created two empty list and I have pass them the first column of each file (AGI-Code column). For this purpouse I have use csv.reader combined with the append command. \n",
    "\n",
    "Once I have the two list containing the AGI codes I have compared them using and if-else statement.\n",
    "\n",
    "The results show that the two list are equal so both files cointain data about the same genes. (AGI-locus code)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connecting to sql server\n",
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating germplasm database\n",
    "%sql create database germplasm;\n",
    "%sql show databases;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Use command its used to select a database and perform SQL operations into that database\n",
    "%sql use germplasm;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating first table (germplasm) and defining columns and different variables\n",
    "%sql CREATE TABLE germplasm (locus VARCHAR(20) PRIMARY KEY, germplasm VARCHAR(50), phenotype VARCHAR(500), pubmed INTEGER(20));\n",
    "%sql DESCRIBE germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating second table (locus_gene) and defining columns and different variables\n",
    "%sql CREATE TABLE locus_gene (locus VARCHAR(20) PRIMARY KEY, gene VARCHAR(20), protein_length INTEGER(10));\n",
    "%sql DESCRIBE locus_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First step has been to connect to sql server. After that I have created the database germplasm and inside of that I have create the two tables, and I have specified the column names and different variables that define the kind of data that can be introduce in each column. Both tables are already linked through the \"Locus\" column, defined as primary key as each entry must have an unique Locus."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasm',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit= False)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('Germplasm.tsv', 'r') as tsvfile:\n",
    "            germ = csv.DictReader(tsvfile, fieldnames= (\"Locus\", \"germplasm\", \"phenotype\",\"pubmed\"),delimiter = \"\\t\")\n",
    "            for row in germ:\n",
    "                print(row)\n",
    "                sql = 'INSERT INTO germplasm (locus, germplasm, phenotype, pubmed) VALUES (\"{}\", \"{}\", \"{}\", {})'.format(row[\"Locus\"], row[\"germplasm\"], row[\"phenotype\"], row[\"pubmed\"])\n",
    "                cursor.execute(sql)\n",
    "                connection.commit()\n",
    "        \n",
    "                \n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv', 'r') as tsvfile:\n",
    "            loc = csv.DictReader(tsvfile, fieldnames= (\"Locus\", \"Gene\", \"ProteinLength\"),delimiter = \"\\t\")\n",
    "            for row in loc:\n",
    "                print(row)\n",
    "                sql = 'INSERT INTO locus_gene (locus, gene, protein_length) VALUES (\"{}\", \"{}\", {})'.format(row[\"Locus\"], row[\"Gene\"], row[\"ProteinLength\"])\n",
    "                cursor.execute(sql)\n",
    "                connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this problem I have used the pymsql cursors to enter the data of the two tsv files in the previously created tables (germplasm and locus_gene).\n",
    "\n",
    "To do this, firstly I have established the connection to the germplasm database using pymsql.connect.\n",
    "\n",
    "Then, using the command \"try\" I have added the necessary code to introduce the data from the file Germplasm.tsv to the corresponding table. First the previously imported csv library is used to read the file and with a for loop the data is entered in the corresponding columns.\n",
    "\n",
    "Finally a print of the previous process is made to check that it has been done correctly.\n",
    "\n",
    "As I have not closed the connection it is not necessary to establish it again so I have directly created the code inside \"try\" for the second table, locus_gene, and this time I have closed the connection."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Report 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasm',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=False)\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql_1 = \"\"\"SELECT germplasm.locus, germplasm, phenotype, gene, protein_length FROM germplasm, locus_gene \n",
    "                WHERE germplasm.locus = locus_gene.locus \"\"\"\n",
    "        cursor.execute(sql_1)\n",
    "        results_1 = cursor.fetchall()\n",
    "\n",
    "finally:\n",
    "    print(results_1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this problem I have used pymsql to do the report. In this case I have had to create a connection using pymsql.connect becouse I have closed it before. I am using two libraries, cvs and pymsql.cursors but I have already imported them in previous cells.\n",
    "\n",
    "Inside try command I have created a cursor that looks for the query specified in the problem, after that, using cursor.fetchall() I have store the result in \"result_1\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Report 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    \n",
    "    with connection.cursor() as cursor:\n",
    "        sql_2 = \"AND gene in ('SKOR', 'MAA3')\"\n",
    "        cursor.execute(sql_1 + sql_2)\n",
    "        results_2 = cursor.fetchall()\n",
    "finally:\n",
    "    print(results_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this problem I have use pymsql cursor to query data about two genes (SKOR and MAA3), to simplify the query I have combined the previous query that join the two tables and I have added a new query (AND) specifying the genes that I wanted to report."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Report 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    },
    {
     "ename": "InterfaceError",
     "evalue": "(0, '')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mInterfaceError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-165-b9f0d57f1abd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mchro\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m6\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m             \u001b[0msql_3\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"SELECT SUBSTRING(locus, 1, 4) AS Prefix, COUNT(locus) AS entries FROM germplasm group by SUBSTRING(locus, 1, 4)\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m             \u001b[0mcursor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msql_3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m             \u001b[0mresults_3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcursor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfetchall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/cursors.py\u001b[0m in \u001b[0;36mexecute\u001b[0;34m(self, query, args)\u001b[0m\n\u001b[1;32m    168\u001b[0m         \u001b[0mquery\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmogrify\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    169\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 170\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_query\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    171\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_executed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mquery\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    172\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/cursors.py\u001b[0m in \u001b[0;36m_query\u001b[0;34m(self, q)\u001b[0m\n\u001b[1;32m    326\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_last_executed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    327\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_clear_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 328\u001b[0;31m         \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    329\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_do_get_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrowcount\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/connections.py\u001b[0m in \u001b[0;36mquery\u001b[0;34m(self, sql, unbuffered)\u001b[0m\n\u001b[1;32m    514\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    515\u001b[0m                 \u001b[0msql\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msql\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'surrogateescape'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 516\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_execute_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCOMMAND\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCOM_QUERY\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msql\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    517\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_affected_rows\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_query_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0munbuffered\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0munbuffered\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    518\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_affected_rows\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/connections.py\u001b[0m in \u001b[0;36m_execute_command\u001b[0;34m(self, command, sql)\u001b[0m\n\u001b[1;32m    748\u001b[0m         \"\"\"\n\u001b[1;32m    749\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 750\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mInterfaceError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"(0, '')\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    751\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    752\u001b[0m         \u001b[0;31m# If the last query was unbuffered, make sure it finishes before\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mInterfaceError\u001b[0m: (0, '')"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    results_3 = []\n",
    "    with connection.cursor() as cursor:\n",
    "        for chro in range(1, 6):\n",
    "            sql_3 = \"SELECT SUBSTRING(locus, 1, 4) AS Prefix, COUNT(locus) AS entries FROM germplasm group by SUBSTRING(locus, 1, 4)\"\n",
    "            cursor.execute(sql_3)\n",
    "            results_3.extend(cursor.fetchall())\n",
    "    \n",
    "finally:\n",
    "    print(results_3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this problem I have use pymsql cursor to query the number of genes presented in each crhomosome accordint to the AGIcodes, to simplify the query I have use an object range that is used to specify to the COUNT argument the differents crhomosomes to query."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Report 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        for i in range(1, 6):\n",
    "            sql_4 = f\"SELECT AVG(protein_length) AS average_protein_length_chromosome{i} FROM locus_gene WHERE locus LIKE 'AT{i}G%'\"\n",
    "            cursor.execute(sql_4)\n",
    "            results_4.extend(cursor.fetchall())\n",
    "finally:\n",
    "    print(results_4)\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this problem I have used a similar approach to the previous one, to simplify the problem I have used a range object, which has allowed me to differentiate the different chromosomes and at the same time calculate with them the average length of the protein. \n",
    "\n",
    "Unlike the previous reports in this one I have already closed the connection because it is the last report. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('report.tsv', writemode) as f:\n",
    "    f.write('Report 1\\n')\n",
    "    for data in results_1:\n",
    "            for val in data.values():\n",
    "                f.write(str(val)) \n",
    "                f.write('\\t') \n",
    "            f.write('\\n\\n')\n",
    "            \n",
    "    f.write('Report 2\\n')\n",
    "    for data in results_2:\n",
    "            for val in data.values():\n",
    "                f.write(str(val)) \n",
    "                f.write('\\t') \n",
    "            f.write('\\n\\n')\n",
    "    f.write('Report 3\\n')\n",
    "    \n",
    "    for data in results_3:\n",
    "            for val in data.items():\n",
    "                f.write(str(val)) \n",
    "                f.write('\\t') \n",
    "            f.write('\\n\\n')\n",
    "    f.write('Report 4\\n')\n",
    "    \n",
    "    for data in results_4:\n",
    "            for val in data.values():\n",
    "                f.write(str(val)) \n",
    "                f.write('\\t') \n",
    "            f.write('\\n\\n')\n",
    "       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With this code I have created a file in tsv format where the different reports made previously."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
